core/network/src/sync.rs

```diff
@@ -496,42 +540,41 @@ impl<B: BlockT> ChainSync<B> {
 						}
 					}).collect()
 				},
-				PeerSyncState::AncestorSearch(n) => {
-					match blocks.get(0) {
-						Some(ref block) => {
-							trace!(target: "sync", "Got ancestry block #{} ({}) from peer {}", n, block.hash, who);
-							match protocol.client().block_hash(n) {
-								Ok(Some(block_hash)) if block_hash == block.hash => {
-									if peer.common_number < n {
-										peer.common_number = n;
-									}
-									peer.state = PeerSyncState::Available;
-									trace!(target:"sync", "Found common ancestor for peer {}: {} ({})", who, block.hash, n);
-									vec![]
-								},
-								Ok(our_best) if n > As::sa(0) => {
-									trace!(target:"sync", "Ancestry block mismatch for peer {}: theirs: {} ({}), ours: {:?}", who, block.hash, n, our_best);
-									let n = n - As::sa(1);
-									peer.state = PeerSyncState::AncestorSearch(n);
-									Self::request_ancestry(protocol, who, n);
-									return None;
-								},
-								Ok(_) => { // genesis mismatch
-									trace!(target:"sync", "Ancestry search: genesis mismatch for peer {}", who);
-									protocol.report_peer(who, Severity::Bad("Ancestry search: genesis mismatch for peer".to_string()));
-									return None;
-								},
-								Err(e) => {
-									let reason = format!("Error answering legitimate blockchain query: {:?}", e);
-									protocol.report_peer(who, Severity::Useless(reason));
-									return None;
-								}
-							}
-						},
-						None => {
+				PeerSyncState::AncestorSearch(num, state) => {
+					let block_hash_match = match (blocks.get(0), protocol.client().block_hash(num)) {
+						(Some(ref block), Ok(Some(our_block_hash))) => block.hash == our_block_hash,
+						(None, _) => {
 							trace!(target:"sync", "Invalid response when searching for ancestor from {}", who);
 							protocol.report_peer(who, Severity::Bad("Invalid response when searching for ancestor".to_string()));
 							return None;
+						},
+						(_, Ok(None)) => {
+							trace!(target:"sync", "None response while trying to get hash of block number: {}", num);
+							return None;
```

Comment:
This should be treated in the same way as mismatched hash. Currently it just leaves peer stuck in `AncestorSearch`.

Reply:
Yeah, good catch.